c8d: Handle userns properly

rumpl opened the pull request:
**- What I did**

Remapped the GID/UID when a snapshot is created if the daemon is started with `--userns-remap`. This implementation differs from the graph driver one slightly. With graph drivers the files/directories are chown'ed when the image is downloaded and the contents extracted. Basically, with graph drivers we remap on pull and with containerd we remap on run.

Note: with this change builds still fail, my guess is because buildkit still mounts the roots in normal mode, I'll continue looking but I think we can merge this one, I am pretty sure some changes in buildkit are needed:

```console
$ docker build -t userns . 
[+] Building 1.7s (6/6) FINISHED                                                                                                                               docker:local
 => [internal] load build definition from dd                                                                                                                           0.0s
 => => transferring dockerfile: 71B                                                                                                                                    0.0s
 => [internal] load .dockerignore                                                                                                                                      0.0s
 => => transferring context: 2B                                                                                                                                        0.0s
 => [internal] load metadata for docker.io/library/debian:bullseye-slim                                                                                                1.3s
 => [auth] library/debian:pull token for registry-1.docker.io                                                                                                          0.0s
 => CACHED [1/2] FROM docker.io/library/debian:bullseye-slim@sha256:61386e11b5256efa33823cbfafd668dd651dbce810b24a8fb7b2e32fa7f65a85                                   0.0s
 => => resolve docker.io/library/debian:bullseye-slim@sha256:61386e11b5256efa33823cbfafd668dd651dbce810b24a8fb7b2e32fa7f65a85                                          0.0s
 => ERROR [2/2] RUN touch testa                                                                                                                                        0.3s
------
 > [2/2] RUN touch testa:
0.287 runc run failed: unable to start container process: error during container init: error mounting "/tmp/root/165536.165536/buildkit/executor/hosts.iv21cdui6qvdwqel9cdo93spo" to rootfs at "/etc/hosts": open /tmp/root/165536.165536/buildkit/executor/7yecjv2i25qdu44khgx41b450/rootfs/etc/hosts: permission denied
------
dd:2
--------------------
   1 |     FROM debian:bullseye-slim
   2 | >>> RUN touch testa
   3 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c touch testa" did not complete successfully: exit code: 1
```

**- How I did it**

**- How to verify it**

`TestDaemonUserNamespaceRootSetting` should not fail any more
 
```console
$ make DOCKER_GRAPHDRIVER=overlayfs TEST_FILTER=TestDaemonUserNamespaceRootSetting TEST_INTEGRATION_USE_SNAPSHOTTER=1 test-integration
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

rumpl committed:
c8d: Handle userns properly

If the daemon is run with --userns-remap we need to chown the prepared
snapshot

Signed-off-by: Djordje Lukic <djordje.lukic@docker.com>

vvoland reviewed (commented):
Does this break `diff`/`commit` (all files are chowned and treated as modified)?

EDIT: Ah, it shouldn't because it creates 2 snapshots on top of the base image.
Wondering if we should also adjust `calculateSnapshotParentUsage` to account for that?

rumpl commented:
> Wondering if we should also adjust `calculateSnapshotParentUsage` to account for that?

Hm... With overlayfs this would mean the reported size is 2x if metacopy is not enabled, which is true since chown would copy the files.

thaJeztah commented:
Looking forward to idmapped mounts, which probably would make these kind of things a lot easier.

dmcgowan commented on daemon/containerd/image_snapshot_unix.go:
```diff
@@ -0,0 +1,70 @@
+		return err
+	}
+
+	if err := remapRootFS(ctx, mounts, rootPair); err != nil {
```
Why does this only use the rootPair and not the full mapping?

rumpl replied:
`RootPair` is the remapped root uid/gid

dmcgowan replied:
Comparing to containerd's implementation which increments over the full userns range https://github.com/containerd/containerd/blob/v1.7.5/container_opts_unix.go#L104

rumpl replied:
Ah right!

rumpl replied:
Fixed thanks

dmcgowan commented on daemon/containerd/image_snapshot_unix.go:
```diff
@@ -0,0 +1,76 @@
+				return fmt.Errorf("cannot get underlying data for %s", path)
+			}
+
+			return os.Lchown(path, int(stat.Uid)+idpair.UID, int(stat.Gid)+idpair.GID)
```
Might be clearer to use `i.idMapping.ToContainer`

rumpl replied:
Ok switched to using `ToHost`, `ToContainer` is the other way around and was giving me errors. With `ToHost` the uid/gid is the right one inside the container. Might be a good idea to add a test for this

rumpl replied:
Added a test for this

vvoland commented on daemon/containerd/image_snapshot_unix.go:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	if err := i.remapRootFS(ctx, mounts); err != nil {
+		snapshotter.Remove(ctx, usernsID) //nolint: errcheck
```
Could we log this error instead of ignoring it?

vvoland commented on daemon/containerd/image_snapshot_windows.go:
```diff
@@ -0,0 +1,14 @@
+//go:build windows
```
Not needed with `_windows` filename suffix

vvoland reviewed (approved):
Overall SGTM; left some nits though

rumpl committed:
c8d/mount: Create same directory owners as graphdrivers

Graph drivers create the parent directory with
rootPair().GID:CurrentIdentity().UID owner. This change brings these in
line

Signed-off-by: Djordje Lukic <djordje.lukic@docker.com>

vvoland commented on daemon/containerd/image_snapshot_unix.go:
```diff
@@ -0,0 +1,85 @@
+	}
+
+	if err := i.remapRootFS(ctx, mounts); err != nil {
+		if rmErr := snapshotter.Remove(ctx, usernsID); err != nil {
```
```suggestion
		if rmErr := snapshotter.Remove(ctx, usernsID); rmErr != nil {
```

thaJeztah commented on daemon/snapshotter/mount.go:
```diff
@@ -93,7 +94,14 @@ func (m *refCountMounter) Mount(mounts []mount.Mount, containerID string) (targe
 
 	root := m.idMap.RootPair()
-	if err := idtools.MkdirAllAndChown(target, 0o700, root); err != nil {
+	dirID := idtools.Identity{
+		UID: idtools.CurrentIdentity().UID,
+		GID: root.GID,
+	}
+	if err := idtools.MkdirAllAndChown(path.Dir(target), 0o710, dirID); err != nil {
+		return "", err
+	}
```
Can you change this to use `filepath.Dir()` ?

I also thought for a long time that `path` is for `/` (and `filepath` only if you need cross-platform); https://github.com/golang/go/blob/2e7e1d2e8d96ddf2aff00feb645cfa94bc9e4786/src/path/path.go#L5-L6

> Package path implements utility routines for manipulating slash-separated paths.

But a second paragraph was added at some point! https://github.com/golang/go/commit/7465bfb1ea1a7ddfec9b267587ee9e6200514f3f

> To manipulate operating system paths, use the path/filepath package.

While at it, perhaps also inline that `dirID` var (I see the other implementations also have the above issues, but we can look at those separately)

```suggestion
	if err := idtools.MkdirAllAndChown(filepath.Dir(target), 0o710, idtools.Identity{
		UID: idtools.CurrentIdentity().UID,
		GID: root.GID,
	}); err != nil {
		return "", err
	}
```

thaJeztah commented on daemon/containerd/image_snapshot_unix.go:
```diff
@@ -0,0 +1,85 @@
+				return fmt.Errorf("cannot get underlying data for %s", path)
+			}
+
+			fid := idtools.Identity{UID: int(stat.Uid), GID: int(stat.Gid)}
+			ids, err := i.idMapping.ToHost(fid)
```
Perhaps inline this one as well;

```suggestion
			ids, err := i.idMapping.ToHost(idtools.Identity{UID: int(stat.Uid), GID: int(stat.Gid)}
```

thaJeztah commented on daemon/snapshotter/mount.go:
```diff
@@ -67,7 +67,6 @@ type refCountMounter struct {
 
 func (m *refCountMounter) Mount(mounts []mount.Mount, containerID string) (target string, retErr error) {
 	target = filepath.Join(m.home, mountsDir, m.snapshotter, containerID)
-
```
unrelated change? 🙈

rumpl replied:
Thought the code looked better that way

thaJeztah reviewed (approved):
LGTM, but spotted some minor issues; let me know if you think it's worth fixing those before merging

rumpl commented on daemon/snapshotter/mount.go:
```diff
@@ -94,11 +93,10 @@ func (m *refCountMounter) Mount(mounts []mount.Mount, containerID string) (targe
 		GID: root.GID,
-	}
-	if err := idtools.MkdirAllAndChown(path.Dir(target), 0o710, dirID); err != nil {
+	}); err != nil {
```
:facepalm: forgot I had two commits in this PR

rumpl replied:
Fixed, thanks

thaJeztah commented on daemon/containerd/image_snapshot.go:
```diff
@@ -63,17 +65,21 @@ func (i *ImageService) PrepareSnapshot(ctx context.Context, id string, parentIma
-	return err
+	snapshotter := i.client.SnapshotService(i.StorageDriver())
+
+	if i.idMapping.Empty() {
+		_, err = snapshotter.Prepare(ctx, id, parentSnapshot)
+		return err
+	}
+
+	return i.remapSnapshot(ctx, snapshotter, id, parentSnapshot, lease)
```
Ohman, changes look good, but I'm just looking at this part again, and this _could be_ a potential foot-gun if either a linter, or some "helpful" contribution fixes "unhandled error" here;

```suggestion

	if i.idMapping.Empty() {
		if _, err := snapshotter.Prepare(ctx, id, parentSnapshot); err != nil {
			return err
		}
	}

	return i.remapSnapshot(ctx, snapshotter, id, parentSnapshot, lease)
```


The above would break this code (which would _hopefully_ be caught in CI.

Do you think it's worth to be defensive here, and to reverse the order?

```suggestion

	if !i.idMapping.Empty() {
		return i.remapSnapshot(ctx, snapshotter, id, parentSnapshot, lease)
	}

	_, err = snapshotter.Prepare(ctx, id, parentSnapshot)
	return err
```

rumpl replied:
Sure I can reverse the order :+1:

thaJeztah reviewed (approved):
LGTM
